libs/vkd3d/memory.c

```diff
@@ -659,9 +762,303 @@ void vkd3d_memory_allocator_cleanup(struct vkd3d_memory_allocator *allocator, st
+    timeline_info.waitSemaphoreValueCount = 1;
+    timeline_info.pWaitSemaphoreValues = &clear_queue->next_signal_value;
+
+    memset(&submit_info, 0, sizeof(submit_info));
+    submit_info.sType = VK_STRUCTURE_TYPE_SUBMIT_INFO;
+    submit_info.pNext = &timeline_info;
+    submit_info.waitSemaphoreCount = 1;
+    submit_info.pWaitSemaphores = &clear_queue->vk_semaphore;
+    submit_info.pWaitDstStageMask = &vk_stage_mask;
+
+    queue_mask = device->unique_queue_mask;
+
+    while (queue_mask)
+    {
+        queue = device->queues[vkd3d_bitmask_iter32(&queue_mask)];
+
+        if (queue == internal_queue)
+            continue;
+
+        if (!(vk_queue = vkd3d_queue_acquire(queue)))
+            return E_FAIL;
+
+        vr = VK_CALL(vkQueueSubmit(vk_queue, 1, &submit_info, VK_NULL_HANDLE));
+        vkd3d_queue_release(queue);
+
+        if (vr < 0)
+        {
+            ERR("Failed to submit semaphore wait, vr %d.\n", vr);
+            return hresult_from_vk_result(vr);
+        }
+    }
+
+    /* Keep next_signal always one ahead of the last signaled value */
+    clear_queue->next_signal_value += 1;
+    clear_queue->num_bytes_pending = 0;
+    clear_queue->allocations_count = 0;
+    clear_queue->command_buffer_index += 1;
+    clear_queue->command_buffer_index %= VKD3D_MEMORY_CLEAR_COMMAND_BUFFER_COUNT;
+    return S_OK;
+}
+
+HRESULT vkd3d_memory_allocator_flush_clears(struct vkd3d_memory_allocator *allocator, struct d3d12_device *device)
+{
+    struct vkd3d_memory_clear_queue *clear_queue = &allocator->clear_queue;
+    HRESULT hr;
+
+    pthread_mutex_lock(&clear_queue->mutex);
+    hr = vkd3d_memory_allocator_flush_clears_locked(allocator, device);
+    pthread_mutex_unlock(&clear_queue->mutex);
+    return hr;
+}
+
+#define VKD3D_MEMORY_CLEAR_QUEUE_MAX_PENDING_BYTES (256ull << 20) /* 256 MiB */
+
+static void vkd3d_memory_allocator_clear_allocation(struct vkd3d_memory_allocator *allocator,
+        struct d3d12_device *device, struct vkd3d_memory_allocation *allocation)
+{
+    struct vkd3d_memory_clear_queue *clear_queue = &allocator->clear_queue;
+
+    if (allocation->cpu_address)
+    {
+        /* Probably faster than doing this on the GPU
+         * and having to worry about synchronization */
+        memset(allocation->cpu_address, 0, allocation->resource.size);
+    }
+    else if (allocation->resource.vk_buffer)
+    {
+        pthread_mutex_lock(&clear_queue->mutex);
+
+        if (!vkd3d_array_reserve((void**)&clear_queue->allocations, &clear_queue->allocations_size,
+                clear_queue->allocations_count + 1, sizeof(*clear_queue->allocations)))
+        {
+            ERR("Failed to insert free range.\n");
+            pthread_mutex_unlock(&clear_queue->mutex);
+            return;
+        }
+
+        allocation->clear_semaphore_value = clear_queue->next_signal_value;
+
+        if (allocation->chunk)
+            allocation->chunk->allocation.clear_semaphore_value = clear_queue->next_signal_value;
```

Comment:
I guess we don't need max here because next_signal_value is monotonically increasing?

Reply:
Yes, and this happens while the mutex is locked, so there can't be any concurrency issues either.